SIMD-0089: Programify Feature Gate Program (#89)
solana-foundation/solana-improvement-documents: proposals/0089-programify-feature-gate-program.md

the case of a mistaken queuing, discovery of a bug, or simply a desire to manage
the cadence and schedule of activations.

A fully-implemented Core BPF program would take ownership of those accounts and
support revoking queued features, giving engineers more flexibility and
safeguards.

## Alternatives Considered

The Feature Gate program could instead be implemented as a built-in native
program, rather than a Core BPF program. However, this would mean any changes to
the program would need to be implemented by all validator clients in
coordination. This makes upgrading the program cumbersome.

With the Feature Gate program instead implemented as a Core BPF program, any
changes need only be done once, eliminating this duplication of work.

## New Terminology

- Feature Gate program: The Core BPF program that all feature accounts will be
  assigned to, with address `Feature111111111111111111111111111111111111`.

## Detailed Design

The Feature Gate program shall be a Core BPF program whose upgrade authority
will be a multi-sig authority, with keyholders from each validator client
implementation. This includes Solana Labs, Jito, and Jump. In the future, this

> can be expanded to include new clients such as Syndica.

Author: Ok so this would just be Solana Labs and Jito. I won't go into specifics on how many keys, etc.

I've updated this in the SIMD to remove Jump.

feature activations.

**Note:** The process by which core contributors *activate* features would
remain completely unchanged.

## Motivation

Currently, a feature is queued for activation by a keypair holder creating an
empty account and assigning it to the
`Feature111111111111111111111111111111111111` program.

Because there is no actual program implementation at this address, the queuing
is irreversible; if the runtime knows about a feature gate at some address, it
will activate it at the next epoch boundary. This means there is no recourse in
the case of a mistaken queuing, discovery of a bug, or simply a desire to manage
the cadence and schedule of activations.

A fully-implemented Core BPF program would take ownership of those accounts and
support revoking queued features, giving engineers more flexibility and
safeguards.

## Alternatives Considered

The Feature Gate program could instead be implemented as a built-in native
program, rather than a Core BPF program. However, this would mean any changes to
the program would need to be implemented by all validator clients in
coordination. This makes upgrading the program cumbersome and potentially

> dangerous.

Reviewer: It's not obvious (to me) how upgrading the program is dangerous, per se. I think I'd just remove that part of this sentence.

> With the Feature Gate program instead implemented as a Core BPF program, the
> program could be changed only once, eliminating this duplication of work.

Reviewer: ```suggestion
With the Feature Gate program instead implemented as a Core BPF program, any changes
need only be done once, eliminating this duplication of work.
```

## New Terminology

- Feature Gate program: The Core BPF program that all feature accounts will be
  assigned to, with address `Feature111111111111111111111111111111111111`.

> - "Revoke" or "revoke pending activation": The act of reallocating a feature
>   account's data to zero, assigning it to the System Program, and defunding its
>   lamports balance - effectively removing it from the runtime's recognized set

Reviewer: ```suggestion
- "Revoke" or "revoke pending activation": The act of removing a feature from the runtime's recognized set
```
These are design details, so just wait and include below?

  of pending feature activations.

## Detailed Design

The Feature Gate program shall be a Core BPF program whose upgrade authority

> will be controlled by a multi-sig authority, with keyholders from each validator

Reviewer: ```suggestion
will be a multi-sig authority, with keyholders from each validator
```

client implementation. This includes Solana Labs, Jito, and Jump. In the future,
this can be expanded to include new clients such as Syndica.

The program shall initially be designed to support one instruction:
`RevokePendingActivation`. Any other instructions or functionality this program

> may support in the future is outside the scope of this SIMD.

Reviewer: ```suggestion
may support in the future are outside the scope of this SIMD.
```
or... "will be proposed and discussed separately"?

> As mentioned above under "New Terminology", when this instruction is invoked by

Reviewer: ```suggestion
When this instruction is invoked by
```

a feature key-holder, the program will reallocate the account to zero, assign it
back to the System Program, and defund its lamports balance. As a result, the
runtime will no longer recognize this feature as pending, since it will no
longer be owned by `Feature111111111111111111111111111111111111`.

Consider the instruction as it may appear in the Feature Gate program:

```rust
pub enum FeatureGateInstruction {
    /// Revoke a pending feature activation.
    ///
    /// A "pending" feature activation is a feature account that has been
    /// allocated and assigned, but hasn't yet been updated by the runtime
    /// with an `activation_slot`.
    ///
    /// Features that _have_ been activated by the runtime cannot be revoked.
    ///
    /// Accounts expected by this instruction:
    ///
    ///   0. `[w+s]`    Feature account
    ///   1. `[w]`      Destination (for rent lamports)
    RevokePendingActivation,
}

> ```

Reviewer: I don't really see much point in this being here. Since we generally try to avoid code in SIMDs, I think just remove it.

Author: Sorry, I thought you guys specifically said to include state and instruction layouts in the SIMDs.

Author: Circling back here, I plucked the code snippet as requested!

> The non-existent program at `Feature111111111111111111111111111111111111` can be
> considered analogous to a no-op native program. Thus, the official processes
> outlined in
> [SIMD 0088](https://github.com/solana-foundation/solana-improvement-documents/pull/88)
> for migrating a native program to Core BPF can be used to enable this new program.
> 
> Consider the following steps to activate Feature Gate:
> 
> 1. Migrate the native no-op program at
>    `Feature111111111111111111111111111111111111` to a Core BPF no-op.
> 2. Upgrade the Core BPF no-op to add the `RevokePendingActivation` instruction.
> 
> Executing these two steps would effectively activate Feature Gate without any
> changes to existing processes.
> 
> As mentioned previously, the upgrade in step 2 would be conducted by the Feature
> Gate program's multi-sig upgrade authority. 

Reviewer: ```suggestion
The official processes
outlined in
[SIMD 0088](https://github.com/solana-foundation/solana-improvement-documents/pull/88)
for migrating a native program to Core BPF will be used to enable this new program, with the addition of the Revoke instruction as a separate BPF rogram-upgrade step.

1. Migrate the native no-op program at
   `Feature111111111111111111111111111111111111` to a Core BPF no-op, with the new program's upgrade authority set to the Feature Gate multisig
2. Upgrade the Core BPF no-op to add the `RevokePendingActivation` instruction using the new multisig authorization

Because the only change to the program is the Revoke addition, these steps will enable BPF Feature Gate without any
changes to existing feature-activation processes.
```
